### ISSUE TITLE
Unable to resolve `@angular/common/locales/*`

### ISSUE THREAD

**Issue by kryshac** [angular-architects/module-federation-plugin#524]
if I work with serve everything works correctly. but after building I get the following error:
```shell
Unable to resolve specifier '@angular/common/locales/en`
```
I think the problem is because of the `*`
![image](https://github.com/angular-architects/module-federation-plugin/assets/19764133/1c113e11-a620-467e-ab7b-6593f3e9efa8)

**Comment by kryshac** [comment]
I added a fix in `@softarc/native-federation` temporarily

```patch
diff --git a/node_modules/@softarc/native-federation/src/lib/utils/package-info.js b/node_modules/@softarc/native-federation/src/lib/utils/package-info.js
index 51680b0..09cd6ee 100644
--- a/node_modules/@softarc/native-federation/src/lib/utils/package-info.js
+++ b/node_modules/@softarc/native-federation/src/lib/utils/package-info.js
@@ -259,6 +259,15 @@ function _getPackageInfo(packageName, directory) {
             esm,
         };
     }
+    cand = secondaryPgkPath + '.mjs';
+    if (fs.existsSync(cand)) {
+        return {
+            entryPoint: cand,
+            packageName,
+            version,
+            esm,
+        };
+    }
     // TODO: Add logger
     logger_1.logger.warn('No entry point found for ' + packageName);
     logger_1.logger.warn("If you don't need this package, skip it in your federation.config.js or consider moving it into depDependencies in your package.json");
```

in `federation.config.js` I added the following configurations:
```js
const { withNativeFederation, shareAll, share } = require('@angular-architects/native-federation/config');

module.exports = withNativeFederation({
  name: 'office',

  shared: {
    ...shareAll({ singleton: true, strictVersion: true, requiredVersion: 'auto' }),

    ...share(
      [
        '@angular/common/locales/en',
        '@angular/common/locales/en-150',
        '@angular/common/locales/en-CA',
        '@angular/common/locales/en-GB',
      ].reduce(
        (acc, name) => ({
          ...acc,
          [name]: { singleton: true, strictVersion: true, requiredVersion: 'auto', includeSecondaries: false },
        }),
        {},
      ),
    ),
  },

  skip: [
    'rxjs/ajax',
    'rxjs/fetch',
    'rxjs/testing',
    'rxjs/webSocket',
    // Add further packages you don't need at runtime
  ],

  sharedMappings: ['@lib/env'],
});
```

I couldn't find where the source from `@softarc/native-federation' is located to see if there is an issue there. @manfredsteyer, I saw that you published, what do you think of such an implementation? (can also be added for .js)

**Comment by digaus** [comment]
Thanks for this issue. I am currently testing out Native Federation for usage in out company and I just could not get it to work with localeData... with this workarround it finally works.

**Comment by Sabbacus** [comment]
I hade this error (or something similar at least). My workaround was to add the locales to the polyfill array.
`"polyfills": ["zone.js", "@angular/localize/init", "@angular/common/locales/global/sv", "@angular/common/locales/global/en-GB"],`

**Comment by kryshac** [comment]
@Sabbacus hello, I also tried your version, but I still get the same error in runtime but also a new one in build.

And I think it's not the best solution because you won't be able to do three shakes, they will all end up in the main bundle

![image](https://github.com/angular-architects/module-federation-plugin/assets/19764133/49148fc5-7681-447a-9584-13c9fa43bf3f)

**Comment by digaus** [comment]
@kryshac  do you have a solution for the mfe ? This workarround makes it work in the wrapper application but when I just open the url of the mfe I still get this error :?

**Comment by kryshac** [comment]
@digaus  did you put this in `mfe/federation.config.js`?
```js
...share(
      [
        '@angular/common/locales/en',
        '@angular/common/locales/en-150',
        '@angular/common/locales/en-CA',
        '@angular/common/locales/en-GB',
      ].reduce(
        (acc, name) => ({
          ...acc,
          [name]: { singleton: true, strictVersion: true, requiredVersion: 'auto', includeSecondaries: false },
        }),
        {},
      ),
    ),
```

**Comment by digaus** [comment]
Yes I did. What's worth the mfe is an angular 16 project.

**Comment by kryshac** [comment]
@digaus can you create a repo with this error because it doesn't reproduce for me?

**Comment by digaus** [comment]
> @digaus can you create a repo with this error because it doesn't reproduce for me?

https://github.com/digaus/mfe1 

Here you go. Opening URL of mfe (http://localhost:4201) results in mentioned error.

Using mfe inside shell application however works.

**Comment by kryshac** [comment]
@digaus hi, I made some changes in this [pull request](https://github.com/digaus/mfe1/pull/1)  and everything works for me as expected.

what changes did i make:

1. I don't know if you had done the package path, I added a script that does this automatically
2. Local de-DE does not exist in Angular

**Comment by digaus** [comment]
Yeah have done it locally but postinstall script is better. Also got it working with my test project. No clue what was wrong before.

Btw do you have an idea for an angular 16 project where I get errors that it cannot import my material styles? Working fine with an angular 17 test project but not with angular 16.

**Comment by kryshac** [comment]
I did not implement the micro front end on version 16, I waited to switch to 17 so that I could directly implement native-federation (because in 16 esbuild was in beta) but what error do you get?

**Comment by digaus** [comment]
Basically this:

https://github.com/angular-architects/module-federation-plugin/issues/373

**Comment by chkhaidzegiorgi** [comment]
i have same issue, actually for me locally everything works fine, locales are loaded, but I have issue when I deploy the project, any suggestions? @kryshac @digaus

**Comment by kryshac** [comment]
@chkhaidzegiorgi the fix is for build/deploy in serve worked without this implementation. what error do you have?

**Comment by chkhaidzegiorgi** [comment]
the error is that - Unable to resolve specifier '@angular/common/locales/de`

@kryshac 
tried to implement the save way as you did with patch but I have same issue

implementation:
![Screenshot 2024-04-02 at 12 52 39](https://github.com/angular-architects/module-federation-plugin/assets/10046491/f71ad93b-6aa0-4abb-ac33-531f18d0ab6b)

**Comment by kryshac** [comment]
@chkhaidzegiorgi  you can look [here](https://github.com/digaus/mfe1/pull/1) and see what is different from yours

**Comment by chkhaidzegiorgi** [comment]
I implemented exactly same way. but the difference is that I am calling registerLocaleData into app.module.ts

**Comment by kryshac** [comment]
@chkhaidzegiorgi create a repo that reproduces the error

**Comment by ips219** [comment]
I'm migrating from webpack to native federation, and I've found same issue. I tried to make the workaround of @kryshac by forcing de missing dependencies in the federation.config.es but it didn't work for me.

For the records... 

I did an small refactor in my code to initialize the list of locales registered by the APP inside the remote module constructor (normally this code is statically initialized outside the module in most of the examples I've seen, and normally ngmodule constructor is not used for anything).

![image](https://github.com/angular-architects/module-federation-plugin/assets/3368591/03d187fc-80dc-4acd-a853-8ed387bf8adc)

In my application this change fixed the issue.

**Comment by chkhaidzegiorgi** [comment]
@kryshac I have solved the issue, my issue was that we were using Docker for deployment, and when Docker was building the image I was not importing patch file.

**Comment by kryshac** [comment]
@chkhaidzegiorgi sorry for the late reply, you need to implement [this](https://www.npmjs.com/package/patch-package), we also use docker for deployment

**Comment by JuNe98** [comment]
@manfredsteyer is there a fix planned for this issue? Because the `patch-package` mentioned by @kryshac  is only a temporary workaround.

**Comment by TimZM** [comment]
I'm also running into this issue and would like to know if this is something that is on the radar @manfredsteyer

**Comment by kudarias** [comment]
My solution is add it locally to project. I know, it is not good approach, but it is locales so i think, they wont change a lot. @manfredsteyer it will be fixed? or not? i know you are busy man, just simple answer, yes or no. :) thnx

**Comment by rainerhahnekamp** [comment]
I'm not Manfred but do I understand it correctly, that you guys @kryshac and @chkhaidzegiorgi have a working and stable solution that "just" needs to be merged?

**Comment by JuNe98** [comment]
@rainerhahnekamp I had the same mentioned issue and the patch mentioned by @kryshac ([Comment](https://github.com/angular-architects/module-federation-plugin/issues/524#issuecomment-2004550256)) works quite well for me.

But I can`t tell what the core problem was here.

**Comment by ianmott77** [comment]
@JuNe98 and @rainerhahnekamp the issue here is that when the entry point is being determined by package-info.js for the shared package it looks for an entry point through either the package.json or it looks for an index.js or index.mjs file however in the case of Angular locales they have no package.json or index file. The locale language files are just an .mjs file with the name of the language for example the language file for **en** is @angular/common/locales/en.mjs. The reason the patch by @kryshac works is that after package-info.js fails to resolve the proper entry point the patch just takes what it most cases would be the parent directory for the package and appends the .mjs extension to it which is the pattern that works in this case, the only other consideration would be if this strategy has any unintended consequences for loading other libraries. 

As a temporary workaround without applying the patch by @kryshac  you can make a new folder in node_modules/@angular/common/locales/ for the language you want and then copy the .mjs for the appropriate language in there and then rename the file index.mjs so that package-info.js  is able to find that as an entry point, you could also do the same thing except instead of copying use symlinks although symlinks won't be as portable. Then you just need to share the language file the same way @kryshac did in your federation.config.js.

Hopefully this is somewhat helpful

**Comment by rainerhahnekamp** [comment]
@ianmott77 thanks for that explanation. I just joined lately here (Native Federation) but I've heard that this is an edge case. It would be great if somebody could come up with a PR.

**Comment by vahidsamimi** [comment]
I have the same issue. How should I fix it? is any update or i must use [https://github.com/digaus/mfe1/pull/1](https://github.com/digaus/mfe1/pull/1)

**Comment by fpd007** [comment]
+1

**Comment by fpd007** [comment]
After spending hours and days to solve this, I finally succeeded by adding this to federation.config.js:
```
shared: {
        ...shareAll({ singleton: true, strictVersion: true, requiredVersion: 'auto' },
            [
                // My skipped packages go here. For some reason it does not workusing the "skip" array (but that is a different issue).
            ]
        ),
        '@angular/common/locales/de': {
            packageInfo: {
                entryPoint: '@angular/common/locales/de',
                esm: true,
            },
        },
    },
```

**Comment by ianmott77** [comment]
@fpd007 yes it seems as of 18.2.7 this is the suggested solution https://github.com/angular-architects/module-federation-plugin/tree/main/libs/native-federation#how-to-manually-define-a-packages-entry-point this issue should be closed now

**Comment by PedroS5499** [comment]
I made this, and solved on my side this issue.
without changed federations source

`TS
...shareAll({ singleton: true, strictVersion: true, requiredVersion: 'auto' }),
		"@angular/common/locales/pt": {
			packageInfo: {
				entryPoint: 'node_modules/@angular/common/locales/pt.mjs',
				singleton: true, strictVersion: true, requiredVersion: 'auto',
				esm: true
			}
		},
		"@angular/common/locales/pt-PT": {
			packageInfo: {
				entryPoint: 'node_modules/@angular/common/locales/pt-PT.mjs',
				singleton: true, strictVersion: true, requiredVersion: 'auto',
				esm: true
			}
		},
`

**Comment by kryshac** [comment]
it was fixed in version `18.2.7`
